### PR TITLE
Backend_Auth

### PR DESCRIPTION
**Pull request by Ytachkin**

### CODE REVIEW

File: iconic/accounts/urls.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from django.urls import include, path
+
+from . import routers, views
+# Хреновина
+urlpatterns = [
+    path('data/', views.UserRetrieveUpdateDestroyAPIView.as_view(),
+         name='user-data'),
+    path('users/', include(routers.router.urls)),
+    ]
```
Comment (Hrustyaschiy) [inline iconic/accounts/urls.py]: Change the comment

File: iconic/accounts/models.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    #birth_date = models.DateField(null=True, blank=True)
+   # pro = models.BooleanField(default=False)
+#
+  #  USERNAME_FIELD = 'email'
+ #   REQUIRED_FIELDS = []
+#
+ #   objects = managers.CustomUserManager()
+#
+   # def __str__(self):
+  #      return f"{self.email}'s custom account"
```
Comment (Hrustyaschiy) [inline iconic/accounts/models.py]: "Душнила" moment, no newline at end of file

File: iconic/accounts/routers.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+from rest_framework import routers
+
+from . import viewsets
+#create an object of the router class and register the viewset
+router = routers.DefaultRouter()
+router.register('', viewsets.CustomUserModelViewSet)
```
Comment (Hrustyaschiy) [inline iconic/accounts/routers.py]: No newline at end of file

File: iconic/accounts/serializers.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+class CustomUserRetrieveSerializer(serializers.ModelSerializer):# To change
+    birth_date = serializers.CharField(required=False)
+    bio = serializers.CharField(required=False)
+    gender = serializers.CharField(required=False)
+
+    class Meta:
+        model = CustomUser
+        fields = ('first_name', 'last_name', 'email',
+                  'bio', 'gender', 'birth_date', 'id')
```
Comment (Hrustyaschiy) [inline iconic/accounts/serializers.py]: No newline at end of file

File: iconic/accounts/urls.py (hunk 1 above)
Comment (Hrustyaschiy) [inline iconic/accounts/urls.py]: No newline at end of file

File: iconic/accounts/views.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+# A view for adding reading or modifying an individual record
+class UserRetrieveUpdateDestroyAPIView(generics.RetrieveUpdateDestroyAPIView):
+    queryset = CustomUser.objects.all()
+    serializer_class = serializers.CustomUserRetrieveSerializer
+    permission_classes = (permissions.IsAuthenticated,)
+
+
+    def get_object(self):
+        return self.request.user
```
Comment (Hrustyaschiy) [inline iconic/accounts/views.py]: No newline at end of file

File: iconic/accounts/viewsets.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+#Let's define a viewset for working with models by copying from the view
+class CustomUserModelViewSet(viewsets.ModelViewSet):
+    serializer_class = serializers.CustomUserSerializer
+    permission_classes = (permissions.AllowAny,)
+    queryset = CustomUser.objects.all()
+
+    def perform_create(self, serializer):
+        instance = serializer.save()
+        instance.set_password(instance.password)
+        instance.save()
```
Comment (Hrustyaschiy) [inline iconic/accounts/viewsets.py]: No newline at end of file

File: iconic/iconic/settings.py [hunk 7]
```diff
@@ -121,3 +121,32 @@
...
+#REST_FRAMEWORK = {
+#    'DEFAULT_AUTHENTICATION_CLASSES': [
+#        'rest_framework.authentication.TokenAuthentication',
+#        'rest_framework.authentication.BasicAuthentication',
+#        'rest_framework.authentication.SessionAuthentication',
+#    ],
+#    'DEFAULT_PERMISSION_CLASSES': [
+#        'rest_framework.permissions.IsAuthenticated',
+#    ]
+#}
```
Comment (Hrustyaschiy) [inline iconic/iconic/settings.py]: No newline at end of file

**Review by Hrustyaschiy — CHANGES_REQUESTED** [review]

File: iconic/accounts/managers.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
+from django.contrib.auth.base_user import BaseUserManager
+from django.utils.translation import gettext_lazy as _
+from validate_email import validate_email
+
+
+class CustomUserManager(BaseUserManager):#we are defining a custom user model that will handle user authentication for us.
```
Comment (Hordd-Scratch) [inline iconic/accounts/managers.py]: The comment should be in a separate line.

File: iconic/accounts/managers.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+            raise ValueError(_('The Email must be set'))
+        email = self.normalize_email(email)
+        if not validate_email(email):
+            raise ValueError(_('Invalid email set'))
+        user = self.model(email=email, **extra_fields)
+        user.set_password(password)
+        user.save()
+        return user
+
+    def create_superuser(self, email, password, **extra_fields): # superuser
```
Comment (Hordd-Scratch) [inline iconic/accounts/managers.py]: The comment should be in a separate line.

File: iconic/accounts/serializers.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+# creating a serializer class
+class CustomUserSerializer(serializers.ModelSerializer):
+    email = serializers.CharField(
+        write_only=True, validators=[validators.UniqueValidator(
+            message='This email already exists',
+            queryset=CustomUser.objects.all()
+        )]
+    )
+    password = serializers.CharField(write_only=True) #here we specified the fields that we will return, and also prescribed the possibilities of using
```
Comment (Hordd-Scratch) [inline iconic/accounts/serializers.py]: The comment should be in a separate line.

File: iconic/accounts/serializers.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    birth_date = serializers.CharField(required=False)
+    bio = serializers.CharField(required=False)
+    gender = serializers.CharField(required=False)
+    last_name = serializers.CharField(required=False)
+    first_name = serializers.CharField(required=False)
+    birth_date = serializers.CharField(required=False)
+
+    class Meta: #fields that we return to the client
```
Comment (Hordd-Scratch) [inline iconic/accounts/serializers.py]: The comment should be in a separate line.

File: iconic/accounts/serializers.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        model = CustomUser
+        fields = ('first_name', 'last_name', 'email',
+                  'password', 'bio', 'gender', 'birth_date')
+
+
+class CustomUserRetrieveSerializer(serializers.ModelSerializer):# To change
```
Comment (Hordd-Scratch) [inline iconic/accounts/serializers.py:30]: The comment should be in a separate line.

File: iconic/iconic/urls.py [hunk 13]
```diff
@@ -19,3 +19,12 @@
 urlpatterns = [
     path('admin/', admin.site.urls),
 ]
+from django.contrib import admin     #for the backend
```
Comment (Hordd-Scratch) [inline iconic/iconic/urls.py:22]: The comment should be in a separate line.

File: iconic/accounts/managers.py (hunk 9 above)
Comment (BarBoley) [reply iconic/accounts/managers.py:18]: It may be worth adding a check for the presence of the email address in the database?

**Review by BarBoley — CHANGES_REQUESTED** [review]
Perform corrections of the points indicated in the comments in the files:
managers.py
models.py
routers.py
serializers.py
urls.py
views.py
viewsets.py
settings.py

**Review by Hrustyaschiy — APPROVED** [review]
Just for the future: change the commit comment before pushing

**Review by BarBoley — APPROVED** [review]

**Review by Hordd-Scratch — APPROVED** [review]

**Pull request merged**